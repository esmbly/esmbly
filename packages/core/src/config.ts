import { RunConfig, Transformer, OutputFormat, Output } from '@esmbly/types';
import * as errors from './errors';

export function validateRunConfig(config: RunConfig): void {
  if (!config || !config.input || !config.transformers || !config.output) {
    throw new Error(errors.MissingConfig());
  }
  if (config.input.length < 1) {
    throw new Error(errors.NoInput());
  }
  if (config.transformers.length < 1) {
    throw new Error(errors.NoTransformers());
  }
  if (config.output.length < 1) {
    throw new Error(errors.NoOutput());
  }
  config.transformers.forEach((transformer: Transformer) => {
    if (typeof transformer.run !== 'function') {
      throw new Error(errors.InvalidTransformer(transformer));
    }
  });
  config.output.forEach((out: Output) => {
    let format = typeof out === 'string' ? out : out.format;
    if (format in OutputFormat === false) {
      throw new Error(errors.InvalidOutput(format));
    }
  });
}
