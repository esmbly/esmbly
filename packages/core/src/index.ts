import { File, RunConfig } from '@esmbly/types';
import { parse } from './parse';
import { validateConfig, validateInputFormat } from './validate';

export async function run(config: RunConfig): Promise<File[]> {
  validateConfig(config);

  let files: File[] = [];
  const trees = parse(config.input, config.transformers[0]);

  for (const transformer of config.transformers) {
    if (transformer.before) {
      await transformer.before();
    }

    validateInputFormat(trees, transformer);

    if (transformer.transform) {
      await transformer.transform(trees);
    }

    if (transformer.createFiles) {
      files = [
        ...files,
        ...(await transformer.createFiles(trees, config.output)),
      ];
    }

    if (transformer.after) {
      await transformer.after();
    }
  }

  return files;
}
