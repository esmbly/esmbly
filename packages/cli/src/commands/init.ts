import { Argv, Arguments } from 'yargs';
import { createConfig } from '../config';

export interface InitOptions {
  default: boolean;
}

export const command = 'init';

export const describe = 'Create an esmbly configuration file';

export const builder = (yargs: Argv): Argv<any> => {
  return yargs
    .version(false)
    .option('default', {
      default: false,
      describe: 'Use default configuration',
      type: 'boolean',
    })
    .help() as Argv;
};

export const handler = async (argv: Arguments & InitOptions): Promise<void> => {
  try {
    const useDefault = argv.default;
    const { fileName, root } = await createConfig(useDefault);
    console.log(`${fileName} created in ${root}`); // TODO: use @esmbly/output
  } catch (err) {
    console.log(err.message || err); // TODO: use @esmbly/output
  }
};
