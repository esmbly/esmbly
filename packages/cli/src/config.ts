import path from 'path';
import stringify from 'stringify-object';
import { Config } from '@esmbly/types';
import {
  writeFile,
  readFile,
  exists,
  getRoot,
  getRelativePathTo,
} from '@esmbly/utils';
import { promptForConfig } from './prompt';

export const DEFAULT_FILE = '.esmblyrc.js';

export async function getTemplateConfig(): Promise<string> {
  const templatePath = path.resolve(__dirname, `../defaults/${DEFAULT_FILE}`);
  const template = await readFile(templatePath);
  return template.toString();
}

export async function getDefaultConfigPath(): Promise<string> {
  const root = await getRoot();
  return path.join(root, DEFAULT_FILE);
}

export async function readConfig(customPath?: string): Promise<Config[]> {
  const configPath = customPath || (await getDefaultConfigPath());
  const config = require(configPath) // eslint-disable-line
  const toArray = (maybeArrayConfig: Config | Config[]): Config[] => {
    if (Array.isArray(maybeArrayConfig)) {
      return maybeArrayConfig;
    }
    return [maybeArrayConfig];
  };
  if (typeof config === 'function') {
    const configFnResult = await config();
    return toArray(configFnResult);
  }
  return toArray(config);
}

export async function createConfig(
  useDefault: boolean,
): Promise<{
  root: string;
  fileName: string;
}> {
  const root = await getRoot();
  const defaultConfigPath = await getDefaultConfigPath();
  if (await exists(defaultConfigPath)) {
    const relativeConfigPath = await getRelativePathTo(defaultConfigPath);
    throw new Error(
      `Config ${relativeConfigPath} already exists. Remove that file first.`,
    );
  } else if (useDefault) {
    const template = await getTemplateConfig();
    await writeFile(defaultConfigPath, template);
  } else {
    const config = await promptForConfig();
    const content = `module.exports = ${stringify(config)}\n`;
    await writeFile(defaultConfigPath, content);
  }
  return { root, fileName: DEFAULT_FILE };
}
