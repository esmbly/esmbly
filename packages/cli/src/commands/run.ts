import { Arguments, Argv } from 'yargs';
import esmbly from '@esmbly/core';
import { Config, Transformer } from '@esmbly/types';
import { outputFactory, readFiles, transformerFactory } from '@esmbly/utils';
import stringify from 'stringify-object';
import output from '@esmbly/output';
import { readConfig } from '../config';

export interface RunOptions {
  config: string;
  silent: boolean;
  input: string[];
  transformers: string[];
  output: string[];
}

export const command = 'run';

export const describe = 'Run esmbly';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const builder = (yargs: Argv): Argv<any> => {
  return yargs
    .version(false)
    .option('config', {
      alias: 'c',
      describe: 'Configuration file path',
      type: 'string',
    })
    .option('silent', {
      default: false,
      describe: 'Silence output to the console',
      type: 'boolean',
    })
    .option('input', {
      alias: 'i',
      describe: 'The files you want to transform',
      type: 'array',
    })
    .option('transformers', {
      alias: 't',
      describe: 'The transformer you want to use',
      type: 'array',
    })
    .option('output', {
      alias: 'o',
      describe: 'The output formats you want to use',
      type: 'array',
    })
    .option('dry-run', {
      describe: 'Run transformations without outputting any files',
      type: 'boolean',
    })
    .option('print-config', {
      describe: 'Print configuration and then abort',
      type: 'boolean',
    })
    .help();
};

export const handler = async (argv: Arguments & RunOptions): Promise<void> => {
  try {
    if (argv.silent) {
      output.setErrorStream(null);
      output.setOutStream(null);
    }

    const config = await readConfig(argv.config);

    if (argv.printConfig) {
      output.out(`${stringify(config, { indent: ' ' })}\n`);
      return;
    }

    const results = await Promise.all(
      config.map(async (c: Config) => {
        const input = await readFiles(argv.input || c.input);
        const transformers = (argv.transformers || c.transformers).map(
          (transformer: string | Transformer) =>
            transformerFactory(transformer),
        );
        if (input.length < 1) {
          throw new Error(`Found 0 files matching pattern: ${c.input}`);
        }
        const output = (argv.output || c.output).map(outputFactory);
        return esmbly.run({ input, output, transformers });
      }),
    );

    if (argv.dryRun) {
      output.out(`${stringify(results, { indent: ' ' })}\n`);
    } else {
      // TODO: write files here
    }
  } catch (err) {
    output.out(`${err.message}\n` || err);
  }
};
