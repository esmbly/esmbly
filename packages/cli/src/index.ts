import yargs, { Arguments } from 'yargs';
import * as init from './commands/init';
import * as run from './commands/run';

export function runCLI(): Arguments {
  return yargs
    .scriptName('esmbly')
    .command(init)
    .command(run)
    .help().argv;
}
