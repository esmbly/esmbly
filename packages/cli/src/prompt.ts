import inquirer from 'inquirer';
import { Config } from '@esmbly/types';
import { getTransformers, getOutputFormats } from '@esmbly/utils';

export async function promptForConfig(): Promise<Config> {
  const { input } = await inquirer.prompt([
    {
      name: 'input',
      message: 'Which files do you want to transform?',
      type: 'input',
    },
  ]);

  const { transformers } = await inquirer.prompt([
    {
      name: 'transformers',
      message: 'Which transformer do you want to use?',
      type: 'checkbox',
      choices: await getTransformers(),
    },
  ]);

  // TODO: Ask to install transformers here?

  const { output } = await inquirer.prompt([
    {
      name: 'output',
      message: 'Which output formats to you want to use?',
      type: 'checkbox',
      choices: await getOutputFormats(transformers),
    },
  ]);

  return { input: [input], transformers, output };
}
