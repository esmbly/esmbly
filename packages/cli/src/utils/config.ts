import path from 'path';
import { Config, InitOptions } from '@esmbly/types';
import {
  exists,
  getRelativePathTo,
  getRoot,
  readFile,
  writeFile,
} from '@esmbly/utils';
import { promptForConfig } from './prompt';
import { installCLI, installTransformers } from './install';

export const DEFAULT_FILE = 'esmbly.config.js';

export async function getTemplateConfig(): Promise<string> {
  const templatePath = path.resolve(
    __dirname,
    `../../defaults/${DEFAULT_FILE}`,
  );
  const template = await readFile(templatePath);
  return template.toString();
}

export function getDefaultConfigPath(): string {
  const root = getRoot();
  return path.join(root, DEFAULT_FILE);
}

type ConfigFnA = () => Config;
type ConfigFnB = () => Config[];
type ConfigFn = ConfigFnA | ConfigFnB;

function toArrayConfig(maybeArrayConfig: Config | Config[]): Config[] {
  if (Array.isArray(maybeArrayConfig)) {
    return maybeArrayConfig;
  }

  return [maybeArrayConfig];
}

export async function readConfig(
  customPath?: string,
  requirer: (requirePath: string) => unknown = require,
): Promise<Config[]> {
  const configPath = customPath || (await getDefaultConfigPath());
  const config = requirer(configPath) as Config | Config[] | ConfigFn;

  if (typeof config === 'function') {
    const configFnResult = await config();
    return toArrayConfig(configFnResult);
  }

  return toArrayConfig(config);
}

export async function createConfig(
  options: InitOptions,
): Promise<{
  root: string;
  fileName: string;
}> {
  const root = getRoot();
  const defaultConfigPath = getDefaultConfigPath();

  if ((await exists(defaultConfigPath)) && !options.force) {
    const relativeConfigPath = getRelativePathTo(defaultConfigPath);
    throw new Error(
      `Config ${relativeConfigPath} already exists. Remove that file first.`,
    );
  } else if (options.default) {
    const template = await getTemplateConfig();
    await writeFile(defaultConfigPath, template, { overwrite: options.force });
  } else {
    const config = await promptForConfig();

    if (config.install) {
      await installTransformers(config.transformers, config.pkgManager);
      await installCLI(config.pkgManager);
    }

    await writeFile(defaultConfigPath, config.module, {
      overwrite: options.force,
    });
  }

  return { fileName: DEFAULT_FILE, root };
}
