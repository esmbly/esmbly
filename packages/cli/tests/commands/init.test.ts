import * as init from '../../src/commands/init';
import * as config from '../../src/config';
import CommandRunner from '../__fixtures__/CommandRunner';

jest.mock('../../src/config');

const command = new CommandRunner(init);

const mockcreateConfig = (): void => {
  (config as any).createConfig = jest.fn();
  (config as any).createConfig.mockResolvedValue({
    fileName: 'file',
    root: 'root',
  });
};

describe('Init', () => {
  it('exposes the correct command', () => {
    expect(init.command).toEqual('init');
  });
  it('exposes the correct description', () => {
    expect(init.describe).toEqual('Create an esmbly configuration file');
  });
  it('exposes the correct help information', async () => {
    const output = await command.run('init --help');
    expect(output).toMatchSnapshot();
  });
  it('calls createConfig to create a config file', async () => {
    mockcreateConfig();
    await command.run('init');
    const [useDefault] = (config.createConfig as any).mock.calls[0];
    expect(useDefault).toEqual(false);
    expect(config.createConfig).toHaveBeenCalledTimes(1);
  });
  it('uses default values when passing the --default flag', async () => {
    mockcreateConfig();
    await command.run('init --default');
    const [useDefault] = (config.createConfig as any).mock.calls[0];
    expect(useDefault).toEqual(true);
    expect(config.createConfig).toHaveBeenCalledTimes(1);
  });
  it.todo('test output by mocking @esmbly/output');
  it.todo('test error output by mocking @esmbly/output');
});
