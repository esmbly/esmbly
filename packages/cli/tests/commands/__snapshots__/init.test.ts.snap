// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Init exposes the correct help information 1`] = `
"processChild.js init

Create an esmbly configuration file

Options:
  --default  Use default configuration                [boolean] [default: false]
  --help     Show help                                                 [boolean]"
`;
